HCP Terraform Docs (#1664)
hashicorp/web-unified-docs: content/terraform-docs-common/docs/cloud-docs/registry/test/dynamic-credentials/gcp.mdx

---
page_title: Use dynamic credentials with the GCP provider in module testing
description: >-
  Configure GCP Workload Identity Federation to trust HCP Terraform's OIDC tokens for module test runs,
  eliminating the need for static GCP credentials.
---

# Configure dynamic credentials for GCP in module testing

@include 'tfc-package-callouts/registry-test-dynamic-credentials.mdx'

You can use HCP Terraform's native OpenID Connect integration with Google Cloud Platform to get dynamic, short-lived credentials for module test runs. This approach uses GCP Workload Identity Federation to authenticate without storing static service account keys.

## Requirements

You need the following to configure dynamic credentials for GCP in module testing:

- A Google Cloud project with permissions to create and configure Workload Identity resources
- An HCP Terraform organization with module testing enabled
- A private registry module with [testing configured](/terraform/cloud-docs/registry/test)
- Permissions in HCP Terraform to modify registry module test settings
- The `gcloud` CLI (optional, for command-line configuration)

## Configure GCP

> ### Create a Workload Identity Pool

Reviewer: ```suggestion
### Create a workload identity pool
```
This phrase is lower case in the GCP docs.

Using the `gcloud` CLI:

```bash
gcloud iam workload-identity-pools providers create-oidc terraform-cloud \
  --location="global" \
  --workload-identity-pool="terraform-module-tests" \
  --issuer-uri="https://app.terraform.io" \
  --allowed-audiences="gcp.workload.identity" \
  --attribute-mapping="google.subject=assertion.sub,attribute.terraform_run_id=assertion.terraform_run_id,attribute.terraform_run_phase=assertion.terraform_run_phase,attribute.terraform_organization_id=assertion.terraform_organization_id,attribute.terraform_organization_name=assertion.terraform_organization_name"
```

### Configure attribute mapping

Attribute mapping extracts claims from the OIDC token and makes them available for authorization. Configure the following mappings:

1. Edit your Workload Identity Provider
2. In the **Attribute Mapping** section, add:
   - `google.subject` → `assertion.sub`
   - `attribute.terraform_run_id` → `assertion.terraform_run_id`
   - `attribute.terraform_run_phase` → `assertion.terraform_run_phase`
   - `attribute.terraform_organization_id` → `assertion.terraform_organization_id`
   - `attribute.terraform_organization_name` → `assertion.terraform_organization_name`

> 3. Click **Save**

Reviewer: ```suggestion
1. Edit your workload identity provider.
1. In the **Attribute Mapping** section, map the following values:
   - Map `google.subject` to `assertion.sub`
   - Map `attribute.terraform_run_id` to `assertion.terraform_run_id`
   - Map `attribute.terraform_run_phase` to `assertion.terraform_run_phase`
   - Map `attribute.terraform_organization_id` to `assertion.terraform_organization_id`
   - Map `attribute.terraform_organization_name` to `assertion.terraform_organization_name`
3. Click **Save**.
```
We should avoid using symbols, emojis, and similar characters unless they appear as part of a UI we're describing.

### Create attribute conditions (optional)

Attribute conditions filter which tokens can authenticate. To allow only module test runs:

1. Edit your Workload Identity Provider
2. In the **Attribute Conditions** section, add:
   ```
   google.subject.matches("organization:.+:module:.+:operation:test_run")
   ```
3. Click **Save**

This ensures only tokens from module test runs (not workspace runs) can use this provider, by matching the subject claim format used for test runs.

### Create a service account

Create a service account that module tests will impersonate:

1. Navigate to **IAM & Admin > Service Accounts**
2. Click **Create Service Account**
3. Enter a name (e.g., `terraform-module-test-sa`)
4. Optionally, add a description
5. Click **Create and Continue**
6. Grant the service account appropriate roles for your tests (see [permissions recommendations](#permissions-recommendations))

> 7. Click **Continue**, then **Done**

Reviewer: ```suggestion
Create a service account for the module tests to impersonate:

1. Navigate to **IAM & Admin > Service Accounts**.
1. Click **Create Service Account**.
1. Enter a name, for example `terraform-module-test-sa`.
1. Optionally, add a description.
1. Click **Create and Continue**.
1. Grant the service account appropriate roles for your tests. Refer to [permissions recommendations](#permissions-recommendations).
1. Click **Continue**, then **Done**.
```

Using the `gcloud` CLI:

```bash
gcloud iam service-accounts create terraform-module-test-sa \
  --display-name="Terraform Module Test Service Account"
```

### Grant Workload Identity Pool access to the service account

> Allow the Workload Identity Pool to impersonate the service account:

Reviewer: ```suggestion
### Grant workload identity pool access to the service account

Complete the following steps to allow the workload identity pool to impersonate the service account:
```

1. Open the service account you created
2. Navigate to the **Permissions** tab
3. Click **Grant Access**
4. In the **New principals** field, enter:
   ```
   principalSet://iam.googleapis.com/projects/PROJECT_NUMBER/locations/global/workloadIdentityPools/terraform-module-tests/*
   ```
   Replace `PROJECT_NUMBER` with your GCP project number.
5. Select the role **Workload Identity User**
6. Click **Save**

Using the `gcloud` CLI:

```bash
gcloud iam service-accounts add-iam-policy-binding terraform-module-test-sa@PROJECT_ID.iam.gserviceaccount.com \
  --role="roles/iam.workloadIdentityUser" \
  --member="principalSet://iam.googleapis.com/projects/PROJECT_NUMBER/locations/global/workloadIdentityPools/terraform-module-tests/*"
```

Replace `PROJECT_ID` with your project ID and `PROJECT_NUMBER` with your project number.

### More specific principal filters

You can create more granular access controls using principal filters:

> **Allow specific organization:**

Reviewer: ```suggestion
The following example grants access to a specific organization:
```

```
principalSet://iam.googleapis.com/projects/PROJECT_NUMBER/locations/global/workloadIdentityPools/terraform-module-tests/attribute.terraform_organization_name/my-org
```

**Allow by subject pattern (using CEL attribute condition on the provider):**

> To filter by module name, use an attribute condition on the Workload Identity Provider:

Reviewer: ```suggestion
You can specify a pattern using the CEL attribute condition on the provider to allow access according to filter by module name. The following example uses an attribute condition on the workload identity provider:
```
I'm not sure if I'm wording that correctly. I'm trying to reword to avoid the double colons and to implement style guidelines about when to use bold and parentheses.

```
google.subject.matches("organization:my-org:module:terraform-gcp-.*:operation:test_run")
```

## Configure HCP Terraform

After configuring GCP, configure your registry module's test configuration to use dynamic credentials.

### Gather required information

> You'll need:

Reviewer: ```suggestion
Collect the following information from GCP:
```

1. **Service account email**: `terraform-module-test-sa@PROJECT_ID.iam.gserviceaccount.com`
2. **Workload provider name**: In the format:
   ```
   projects/PROJECT_NUMBER/locations/global/workloadIdentityPools/terraform-module-tests/providers/terraform-cloud

>    ```

Reviewer: ````suggestion
1. **Service account email**: `terraform-module-test-sa@PROJECT_ID.iam.gserviceaccount.com`
1. **Workload provider name** in the following format:

   ```
   projects/PROJECT_NUMBER/locations/global/workloadIdentityPools/terraform-module-tests/providers/terraform-cloud
   ```
````

  --role="roles/viewer"
```

### Resource creation for testing

For modules that create test resources, grant specific roles:

**For Compute Engine resources:**
```bash
gcloud projects add-iam-policy-binding PROJECT_ID \
  --member="serviceAccount:terraform-module-test-sa@PROJECT_ID.iam.gserviceaccount.com" \
  --role="roles/compute.instanceAdmin.v1"
```

**For Cloud Storage:**
```bash
gcloud projects add-iam-policy-binding PROJECT_ID \
  --member="serviceAccount:terraform-module-test-sa@PROJECT_ID.iam.gserviceaccount.com" \
  --role="roles/storage.admin"
```

**For multiple services:**
Consider creating a custom role with only the necessary permissions:

```bash
gcloud iam roles create moduleTestRole --project=PROJECT_ID \
  --title="Module Test Role" \
  --description="Custom role for module testing" \
  --permissions="compute.instances.create,compute.instances.delete,storage.buckets.create,storage.buckets.delete"

gcloud projects add-iam-policy-binding PROJECT_ID \
  --member="serviceAccount:terraform-module-test-sa@PROJECT_ID.iam.gserviceaccount.com" \
  --role="projects/PROJECT_ID/roles/moduleTestRole"
```

> -> **Tip**: Use resource labels like `testing: true` or `managed-by: terraform-test` to identify and manage test resources. You can then use IAM conditions to restrict the service account to only resources with those labels.

Reviewer: ```suggestion
<Tip>

Use resource labels, such as `testing: true` or `managed-by: terraform-test`, to identify and manage test resources. You can then use IAM conditions to restrict the service account to only resources with those labels.

</Tip>
```

> ## Verify the configuration

Reviewer: Move so that it comes immediately after the Configure HCP Terraform section.

After configuring both GCP and HCP Terraform, verify the setup by running a test:

1. Navigate to your module in HCP Terraform
2. Go to the **Tests** tab
3. Click **Start Test Run** or trigger a test through your VCS
4. Monitor the test run logs for successful authentication

If authentication succeeds, you'll see log output indicating the GCP provider authenticated using the OIDC token. If it fails, check the [troubleshooting](#troubleshooting) section.

## Troubleshooting

### Error: "Unable to impersonate service account"

This error indicates the Workload Identity Pool cannot impersonate the service account. Check:

1. The service account exists in your GCP project
2. The Workload Identity User role is granted to the correct principal set
3. The principal set includes the correct project number and pool name

> 4. Attribute conditions in the Workload Identity Provider match your token claims

Reviewer: ```suggestion
1. The service account exists in your GCP project.
1. The workload identity user role is granted to the correct principal set.
1. The principal set includes the correct project number and pool name.
1. Attribute conditions in the workload identity provider match your token claims.
```

### Error: "Invalid audience"

The token's audience doesn't match the Workload Identity Provider configuration. Verify:

1. The Workload Identity Provider's allowed audiences includes `gcp.workload.identity`
2. The test configuration specifies the same audience

> 3. Custom audiences match on both sides

Reviewer: ```suggestion
The token's audience doesn't match the workload identity provider configuration. Verify the following conditions:

1. The workload identity provider's allowed audiences includes `gcp.workload.identity`.
1. The test configuration specifies the same audience.
1. Custom audiences match on both sides.
```

### Error: "Token validation failed"

The Workload Identity Provider rejected the token. Check:

1. The issuer URL matches exactly (app.terraform.io or your TFE hostname)
2. The issuer URL doesn't include a trailing slash
3. Attribute mappings are configured correctly
4. Attribute conditions allow the token (e.g., subject matches the expected pattern)

### Error: "Permission denied" or "Forbidden"

The service account lacks necessary permissions. Verify:

1. The service account has appropriate IAM roles granted
2. Organization policies aren't blocking the service account
3. Resource-level IAM permissions allow the service account
4. The service account is enabled (not disabled)

### Debugging principal filters

To debug which principal would be authenticated, check the token's subject claim format. For module tests, it's:

```
organization:my-org:module:terraform-gcp-vpc:operation:test_run
```

Your principal filter must match this format. Common issues:

- Using `principal://` instead of `principalSet://` (the latter allows wildcards)
- Incorrect project number (not project ID)
- Typos in pool or provider names

> - Missing attribute mappings required by your filter

Reviewer: ```suggestion
To debug which principal would is authenticated, verify that the token's subject claim matches the following format for module tests:

```
organization:my-org:module:terraform-gcp-vpc:operation:test_run
```

Your principal filter must match this format. Common issues include the following conditions:

- Using `principal://` instead of `principalSet://`. The latter allows wildcards.
- Incorrect project number and not project ID.
- Typos in pool or provider names.
- Missing attribute mappings required by your filter.
```

### View authentication logs

Enable Cloud Audit Logs for your Workload Identity Pool to see authentication attempts:

1. Navigate to **IAM & Admin > Audit Logs**
2. Find **IAM Service** and **Workload Identity Pools**
3. Enable **Admin Read**, **Data Read**, and **Data Write**
4. Check logs in **Logging > Logs Explorer** with the filter:
   ```
   resource.type="iam_workload_identity_pool"

>    ```

Reviewer: ```suggestion
Enable Cloud Audit Logs for your workload identity pool to see authentication attempts:

1. Navigate to **IAM & Admin > Audit Logs**
1. Find **IAM Service** and **Workload Identity Pools**
1. Enable **Admin Read**, **Data Read**, and **Data Write**
1. Check logs in **Logging > Logs Explorer** with the filter:

   ```
   resource.type="iam_workload_identity_pool"
   ```
```
